readthedocs/proxito/views/mixins.py

```diff
@@ -30,52 +36,129 @@ class ServeDocsMixin:
+
+        :param check_if_exists: If `True` we check if the file exists before trying
+         to serve it.
+        """
+        base_storage_path = project.get_storage_path(
+            type_=MEDIA_TYPE_HTML,
+            version_slug=version.slug,
+            include_file=False,
+            version_type=version.type,
+        )
+
+        # Handle our backend storage not supporting directory indexes,
+        # so we need to append index.html when appropriate.
+        if not filename or filename.endswith("/"):
+            filename += "index.html"
+
+        # If the filename starts with `/`, the join will fail,
+        # so we strip it before joining it.
+        storage_path = build_media_storage.join(base_storage_path, filename.lstrip("/"))
+
+        if check_if_exists and not build_media_storage.exists(storage_path):
+            return None
+
+        self._track_pageview(
+            project=project,
+            path=filename,
+            request=request,
             download=False,
-            version_slug=None,
-    ):
+        )
+        return self._serve_file(
+            request=request,
+            storage_path=storage_path,
+            storage_backend=build_media_storage,
+        )
+
+    def _serve_dowload(self, request, project, version, type_):
         """
-        Serve documentation in the way specified by settings.
+        Serve downloadable content for the given version.
+
+        The HTTP header ``Content-Disposition`` is added with the proper
+        filename (e.g. "pip-pypa-io-en-latest.pdf" or "pip-pypi-io-en-v2.0.pdf"
+        or "docs-celeryproject-org-kombu-en-stable.pdf").
+        """
+        storage_path = project.get_storage_path(
+            type_=type_,
+            version_slug=version.slug,
+            version_type=version.type,
+            include_file=True,
+        )
+        self._track_pageview(
+            project=project,
+            path=storage_path,
+            request=request,
+            download=True,
+        )
+        response = self._serve_file(
+            request=request,
+            storage_path=storage_path,
+            storage_backend=build_media_storage,
+        )
+
+        # Set the filename of the download.
+        filename_ext = storage_path.rsplit(".", 1)[-1]
+        domain = unicode_slugify(project.subdomain().replace(".", "-"))
+        if project.is_subproject:
+            filename = f"{domain}-{project.alias}-{project.language}-{version.slug}.{filename_ext}"
+        else:
+            filename = f"{domain}-{project.language}-{version.slug}.{filename_ext}"
+        response["Content-Disposition"] = f"filename={filename}"
+        return response
+
+    def _serve_file(self, request, storage_path, storage_backend):
+        """
+        Serve a file from storage.
 
         Serve from the filesystem if using ``PYTHON_MEDIA``. We definitely
         shouldn't do this in production, but I don't want to force a check for
         ``DEBUG``.
```

Comment:
We should probably kill `PYTHON_MEDIA` now that we have all the set ups using Docker with NGINX 😄